planning/mission_planner/src/rerouting_static_obstacle/rerouting_static_obstacle.cpp

```diff
@@ -0,0 +1,203 @@
+    lanelet::routing::LaneletPath alternative_route_lanelets;
+    auto alternative_route_found =
+      search_alternative_route(selected_point_lanelet, alternative_route_lanelets);
+
+    if (alternative_route_found) {
+      change_route(alternative_route_lanelets);
+    }
+  }
+}
+
+bool ReroutingStaticObstacle::get_selected_point_lanelets(
+  const geometry_msgs::msg::Point & selected_point,
+  lanelet::ConstLanelets & selected_point_lanelets) const
+{
+  if (!lanelet::utils::query::getCurrentLanelets(
+        road_lanelets_, selected_point, &selected_point_lanelets)) {
+    RCLCPP_WARN_STREAM(this->get_logger(), "Failed to find selected_point current lanelet.");
+    return false;
+  }
+
+  return true;
+}
+
+bool ReroutingStaticObstacle::get_remaining_route_lanelets(
+  lanelet::ConstLanelets & remaining_route_lanelets) const
+{
+  return route_handler_.planPathLaneletsBetweenCheckpoints(
+    current_pose_, goal_pose_, &remaining_route_lanelets);
+}
+
+bool ReroutingStaticObstacle::is_selected_point_in_route(
+  const lanelet::ConstLanelets & selected_point_lanelets,
+  const lanelet::ConstLanelets & remaining_route_lanelets,
+  lanelet::ConstLanelet & selected_point_lanelet) const
+{
+  for (const auto & slc_pnt_llt : selected_point_lanelets) {
+    for (const auto & rmn_rt_llt : remaining_route_lanelets) {
+      if (slc_pnt_llt.id() == rmn_rt_llt.id()) {
+        selected_point_lanelet = slc_pnt_llt;
+        return true;
+      }
+    }
+  }
+
+  return false;
+}
+
+bool ReroutingStaticObstacle::search_alternative_route(
+  const lanelet::ConstLanelet & selected_point_lanelet,
+  lanelet::routing::LaneletPath & alternative_route_lanelets) const
+{
+  bool alternative_route_found{false};
+
+  lanelet::Optional<lanelet::routing::Route> optional_route;
+
+  lanelet::ConstLanelet current_lanelet;
+  double alternative_route_length2d = std::numeric_limits<double>::max();
+  if (!lanelet::utils::query::getClosestLanelet(road_lanelets_, current_pose_, &current_lanelet)) {
+    RCLCPP_WARN_STREAM(this->get_logger(), "Failed to find current_lanelet.");
+    return false;
+  }
+
+  lanelet::ConstLanelet goal_lanelet;
+  if (!lanelet::utils::query::getClosestLanelet(road_lanelets_, goal_pose_, &goal_lanelet)) {
+    RCLCPP_WARN_STREAM(this->get_logger(), "Failed to find goal_lanelet.");
+    return false;
+  }
+  for (const auto & llt : road_lanelets_) {
+    lanelet::ConstLanelets via_lanelet;
+    via_lanelet.push_back(llt);
+    optional_route = routing_graph_ptr_->getRouteVia(current_lanelet, via_lanelet, goal_lanelet, 0);
+
+    if (
+      (optional_route) && (!optional_route->contains(selected_point_lanelet)) &&
+      (optional_route->length2d() < alternative_route_length2d)) {
+      alternative_route_length2d = optional_route->length2d();
+      alternative_route_lanelets = optional_route->shortestPath();
+      alternative_route_found = true;
+    }
+    via_lanelet.clear();
+  }
+
+  return alternative_route_found;
```

Comment:
The loop for all lanes here seems to have a high computational cost, but is this a problem?

What do you think about re-setting road_lanes without blocked lanes in routing_graph, or using `routing_graph_ptr_->following` to find candidate lanes by looking for branch points from blocked lanes?

And I asked in the Lanelet2 issue if there is a better way.
https://github.com/fzi-forschungszentrum-informatik/Lanelet2/issues/334

Reply:
Once the current implementation is OK with your use cases, I'm OK with it.

Reply:
I don't like the idea of finding all possible routes via all road_lanelets_, but in general this seems to be a difficult problem.
Instead, I suggest creating a temporal LaneletMap excluding the `selected_point_lanelets`, re-build the RoutingGraph again and then find the shortestPath on the graph.

Reply:
Generally Dijkstra algorithm's order is O ((V+E)logV), so if we query all road_lanelets to find the minimum cost path, it requires O(V(V+E)logV) which is too heavy. If we reconstruct RoutingGraph and then find the shortest path on it, we would not need such computation cost at least

Reply:
I just noticed we already had similar function in route_handler.
https://github.com/tier4/autoware.universe/blob/b2cd2aac9ebd5bb3e8468c10ee2b008a6ef1101b/planning/route_handler/src/route_handler.cpp#L2249-L2273
We may need to use more efficient approach in the future, but as of now it's OK.